File: src/main/java/org/example/kalkulationsprogramm/controller/SystemSettingsController.java
package org.example.kalkulationsprogramm.controller;

import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;

import org.example.kalkulationsprogramm.service.SystemSettingsService;
import org.example.kalkulationsprogramm.service.SystemSettingsService.TestResult;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import lombok.RequiredArgsConstructor;

/**
 * REST-Controller für System-Einstellungen (API Keys, SMTP, etc.).
 * Nur für authentifizierte Admins erreichbar.
 */
@RestController
@RequestMapping("/api/settings")
@RequiredArgsConstructor
public class SystemSettingsController {

    private final SystemSettingsService settingsService;

    // Pragmatischer E-Mail-Regex – bewusst keine RFC-5322-Voll-Compliance, sondern
    // genau das was Anwender erwarten: nicht-leerer Local-Part, "@",
    // nicht-leerer Domain-Part mit mindestens einem Punkt, alles ohne Whitespace.
    // Ungültiges wie "@", "a@", " @ " wird abgewiesen.
    private static final Pattern EMAIL_PATTERN =
            Pattern.compile("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$");

    // ==================== Alle Einstellungen lesen ====================

    @GetMapping
    public ResponseEntity<Map<String, String>> getAll() {
        return ResponseEntity.ok(settingsService.getAllSettings());
    }

    // ==================== SMTP ====================

    @GetMapping("/smtp")
    public ResponseEntity<SmtpSettingsResponse> getSmtp() {
        return ResponseEntity.ok(new SmtpSettingsResponse(
                settingsService.getSmtpHost(),
                settingsService.getSmtpPort(),
                settingsService.getSmtpUsername(),
                hasValue(settingsService.getSmtpPassword())
        ));
    }

    @PutMapping("/smtp")
    public ResponseEntity<Map<String, String>> saveSmtp(@RequestBody SmtpSettingsRequest req) {
        if (!hasValue(req.host())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Bitte einen gültigen SMTP-Server eintragen."));
        }

        if (!hasValue(req.username())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Bitte einen gültigen SMTP-Benutzernamen eintragen."));
        }

        String effectivePassword = req.password();
        if (effectivePassword == null || effectivePassword.isBlank()) {
            effectivePassword = settingsService.getSmtpPassword();
        }

        settingsService.saveSmtpSettings(req.host(), req.port(), req.username(), effectivePassword);
        return ResponseEntity.ok(Map.of("message", "SMTP-Einstellungen gespeichert."));
    }

    @PostMapping("/smtp/test")
    public ResponseEntity<TestResult> testSmtp(@RequestBody SmtpTestRequest req) {
        String host = req.host() != null ? req.host() : settingsService.getSmtpHost();
        int port = req.port() > 0 ? req.port() : settingsService.getSmtpPort();
        String username = req.username() != null ? req.username() : settingsService.getSmtpUsername();
        String password = req.password() != null ? req.password() : settingsService.getSmtpPassword();

        TestResult result = settingsService.testSmtp(host, port, username, password, req.testRecipient());
        return ResponseEntity.ok(result);
    }

    // ==================== IMAP ====================

    @GetMapping("/imap")
    public ResponseEntity<ImapSettingsResponse> getImap() {
        return ResponseEntity.ok(new ImapSettingsResponse(
                settingsService.getImapHost(),
                settingsService.getImapPort(),
                settingsService.getImapUsername(),
                hasValue(settingsService.getImapPassword())
        ));
    }

    @PutMapping("/imap")
    public ResponseEntity<Map<String, String>> saveImap(@RequestBody ImapSettingsRequest req) {
        if (!hasValue(req.host())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Bitte einen gültigen IMAP-Server eintragen."));
        }
        if (!hasValue(req.username())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Bitte einen gültigen IMAP-Benutzernamen eintragen."));
        }

        String effectivePassword = req.password();
        if (effectivePassword == null || effectivePassword.isBlank()) {
            effectivePassword = settingsService.getImapPassword();
        }

        int port = req.port() > 0 ? req.port() : 993;
        settingsService.saveImapSettings(req.host(), port, req.username(), effectivePassword);
        return ResponseEntity.ok(Map.of("message", "IMAP-Einstellungen gespeichert."));
    }

    @PostMapping("/imap/test")
    public ResponseEntity<TestResult> testImap(@RequestBody ImapTestRequest req) {
        String host = hasValue(req.host()) ? req.host() : settingsService.getImapHost();
        int port = req.port() > 0 ? req.port() : settingsService.getImapPort();
        String username = hasValue(req.username()) ? req.username() : settingsService.getImapUsername();
        String password = (req.password() != null && !req.password().isBlank())
                ? req.password() : settingsService.getImapPassword();

        TestResult result = settingsService.testImap(host, port, username, password);
        return ResponseEntity.ok(result);
    }

    // ==================== Kombiniertes E-Mail-Konto (Einfache Einrichtung) ====================

    /**
     * Speichert E-Mail + Passwort einmalig für SMTP und IMAP. Hosts/Ports bleiben
     * unverändert (Defaults bzw. das, was unter "Erweitert" gesetzt wurde).
     */
    @PutMapping("/email-account")
    public ResponseEntity<Map<String, String>> saveEmailAccount(@RequestBody EmailAccountRequest req) {
        if (!hasValue(req.email())) {
            return ResponseEntity.badRequest().body(Map.of("message", "Bitte eine gültige E-Mail-Adresse eintragen."));
        }

        settingsService.saveEmailAccount(req.email(), req.password());
        return ResponseEntity.ok(Map.of("message", "E-Mail-Konto gespeichert."));
    }

    // ==================== Gemini API Key ====================

    @GetMapping("/gemini")
    public ResponseEntity<GeminiSettingsResponse> getGemini() {
        return ResponseEntity.ok(new GeminiSettingsResponse(
                hasValue(settingsService.getGeminiApiKey())
        ));
    }

    @PutMapping("/gemini")
    public ResponseEntity<Map<String, String>> saveGemini(@RequestBody GeminiSettingsRequest req) {
        settingsService.saveGeminiApiKey(req.apiKey());
        return ResponseEntity.ok(Map.of("message", "Gemini API Key gespeichert."));
    }

    @PostMapping("/gemini/test")
    public ResponseEntity<TestResult> testGemini(@RequestBody GeminiTestRequest req) {
        String apiKey = req.apiKey() != null ? req.apiKey() : settingsService.getGeminiApiKey();
        TestResult result = settingsService.testGeminiApiKey(apiKey);
        return ResponseEntity.ok(result);
    }

    // ==================== Standard-Absender für Auto-Mails ====================

    @GetMapping("/mail-from")
    public ResponseEntity<MailFromResponse> getMailFrom() {
        return ResponseEntity.ok(new MailFromResponse(
                settingsService.getMailFromAddress(),
                settingsService.getSmtpUsername()));
    }

    @PutMapping("/mail-from")
    public ResponseEntity<Map<String, String>> saveMailFrom(@RequestBody MailFromRequest req) {
        String address = req.address() == null ? "" : req.address().trim();
        if (!address.isBlank() && !EMAIL_PATTERN.matcher(address).matches()) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Bitte eine gültige E-Mail-Adresse eintragen."));
        }
        settingsService.saveMailFromAddress(address);
        return ResponseEntity.ok(Map.of("message", address.isBlank()
                ? "Absender zurückgesetzt – Auto-Mails nutzen wieder den SMTP-Benutzer."
                : "Absender für automatische Mails gespeichert."));
    }

    // ==================== Funnel-Spam-Filter ====================

    @GetMapping("/anfrage-funnel-spamfilter")
    public ResponseEntity<FunnelSpamFilterResponse> getFunnelSpamFilter() {
        return ResponseEntity.ok(new FunnelSpamFilterResponse(
                settingsService.isAnfrageFunnelSpamFilterAktiv()));
    }

    @PutMapping("/anfrage-funnel-spamfilter")
    public ResponseEntity<Map<String, String>> saveFunnelSpamFilter(
            @RequestBody FunnelSpamFilterRequest req) {
        settingsService.saveAnfrageFunnelSpamFilterAktiv(req.aktiv());
        return ResponseEntity.ok(Map.of(
                "message", req.aktiv()
                        ? "Spam-Filter für Webseiten-Anfragen aktiviert."
                        : "Spam-Filter für Webseiten-Anfragen deaktiviert."));
    }

    // ==================== DTOs ====================

    private boolean hasValue(String val) {
        if (val == null) {
            return false;
        }

        String normalized = val.trim().toLowerCase(Locale.ROOT);
        return !normalized.isBlank()
                && !"override_in_local".equals(normalized)
                && !"smtp.example.com".equals(normalized)
                && !"change_me_strong_password".equals(normalized);
    }

    record SmtpSettingsResponse(String host, int port, String username, boolean passwordSet) {}
    record SmtpSettingsRequest(String host, int port, String username, String password) {}
    record SmtpTestRequest(String host, int port, String username, String password, String testRecipient) {}
    record ImapSettingsResponse(String host, int port, String username, boolean passwordSet) {}
    record ImapSettingsRequest(String host, int port, String username, String password) {}
    record ImapTestRequest(String host, int port, String username, String password) {}
    record EmailAccountRequest(String email, String password) {}
    record GeminiSettingsResponse(boolean apiKeySet) {}
    record GeminiSettingsRequest(String apiKey) {}
    record GeminiTestRequest(String apiKey) {}
    record FunnelSpamFilterResponse(boolean aktiv) {}
    record FunnelSpamFilterRequest(boolean aktiv) {}
    record MailFromResponse(String address, String smtpUsername) {}
    record MailFromRequest(String address) {}
}

File: src/main/kotlin/org/example/kalkulationsprogramm/controller/SystemSettingsController.kt
package org.example.kalkulationsprogramm.controller

import org.example.kalkulationsprogramm.service.SystemSettingsService
import org.example.kalkulationsprogramm.service.SystemSettingsService.TestResult
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.Locale
import java.util.regex.Pattern

@RestController
@RequestMapping("/api/settings")
class SystemSettingsController(
    private val settingsService: SystemSettingsService,
) {
    @GetMapping
    fun getAll(): ResponseEntity<Map<String, String>> {
        return ResponseEntity.ok(settingsService.allSettings)
    }

    @GetMapping("/smtp")
    fun getSmtp(): ResponseEntity<SmtpSettingsResponse> {
        return ResponseEntity.ok(
            SmtpSettingsResponse(
                settingsService.smtpHost,
                settingsService.smtpPort,
                settingsService.smtpUsername,
                hasValue(settingsService.smtpPassword),
            ),
        )
    }

    @PutMapping("/smtp")
    fun saveSmtp(@RequestBody req: SmtpSettingsRequest): ResponseEntity<Map<String, String>> {
        if (!hasValue(req.host)) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte einen gültigen SMTP-Server eintragen."))
        }
        if (!hasValue(req.username)) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte einen gültigen SMTP-Benutzernamen eintragen."))
        }

        val effectivePassword = req.password?.takeUnless { it.isBlank() } ?: settingsService.smtpPassword
        settingsService.saveSmtpSettings(req.host, req.port, req.username, effectivePassword)
        return ResponseEntity.ok(mapOf("message" to "SMTP-Einstellungen gespeichert."))
    }

    @PostMapping("/smtp/test")
    fun testSmtp(@RequestBody req: SmtpTestRequest): ResponseEntity<TestResult> {
        val host = req.host ?: settingsService.smtpHost
        val port = if (req.port > 0) req.port else settingsService.smtpPort
        val username = req.username ?: settingsService.smtpUsername
        val password = req.password ?: settingsService.smtpPassword
        return ResponseEntity.ok(settingsService.testSmtp(host, port, username, password, req.testRecipient))
    }

    @GetMapping("/imap")
    fun getImap(): ResponseEntity<ImapSettingsResponse> {
        return ResponseEntity.ok(
            ImapSettingsResponse(
                settingsService.imapHost,
                settingsService.imapPort,
                settingsService.imapUsername,
                hasValue(settingsService.imapPassword),
            ),
        )
    }

    @PutMapping("/imap")
    fun saveImap(@RequestBody req: ImapSettingsRequest): ResponseEntity<Map<String, String>> {
        if (!hasValue(req.host)) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte einen gültigen IMAP-Server eintragen."))
        }
        if (!hasValue(req.username)) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte einen gültigen IMAP-Benutzernamen eintragen."))
        }

        val effectivePassword = req.password?.takeUnless { it.isBlank() } ?: settingsService.imapPassword
        val port = if (req.port > 0) req.port else 993
        settingsService.saveImapSettings(req.host, port, req.username, effectivePassword)
        return ResponseEntity.ok(mapOf("message" to "IMAP-Einstellungen gespeichert."))
    }

    @PostMapping("/imap/test")
    fun testImap(@RequestBody req: ImapTestRequest): ResponseEntity<TestResult> {
        val host = if (hasValue(req.host)) req.host else settingsService.imapHost
        val port = if (req.port > 0) req.port else settingsService.imapPort
        val username = if (hasValue(req.username)) req.username else settingsService.imapUsername
        val password = if (!req.password.isNullOrBlank()) req.password else settingsService.imapPassword
        return ResponseEntity.ok(settingsService.testImap(host, port, username, password))
    }

    @PutMapping("/email-account")
    fun saveEmailAccount(@RequestBody req: EmailAccountRequest): ResponseEntity<Map<String, String>> {
        if (!hasValue(req.email)) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte eine gültige E-Mail-Adresse eintragen."))
        }
        settingsService.saveEmailAccount(req.email, req.password)
        return ResponseEntity.ok(mapOf("message" to "E-Mail-Konto gespeichert."))
    }

    @GetMapping("/gemini")
    fun getGemini(): ResponseEntity<GeminiSettingsResponse> {
        return ResponseEntity.ok(GeminiSettingsResponse(hasValue(settingsService.geminiApiKey)))
    }

    @PutMapping("/gemini")
    fun saveGemini(@RequestBody req: GeminiSettingsRequest): ResponseEntity<Map<String, String>> {
        settingsService.saveGeminiApiKey(req.apiKey)
        return ResponseEntity.ok(mapOf("message" to "Gemini API Key gespeichert."))
    }

    @PostMapping("/gemini/test")
    fun testGemini(@RequestBody req: GeminiTestRequest): ResponseEntity<TestResult> {
        val apiKey = req.apiKey ?: settingsService.geminiApiKey
        return ResponseEntity.ok(settingsService.testGeminiApiKey(apiKey))
    }

    @GetMapping("/mail-from")
    fun getMailFrom(): ResponseEntity<MailFromResponse> {
        return ResponseEntity.ok(MailFromResponse(settingsService.mailFromAddress, settingsService.smtpUsername))
    }

    @PutMapping("/mail-from")
    fun saveMailFrom(@RequestBody req: MailFromRequest): ResponseEntity<Map<String, String>> {
        val address = req.address?.trim() ?: ""
        if (address.isNotBlank() && !EMAIL_PATTERN.matcher(address).matches()) {
            return ResponseEntity.badRequest().body(mapOf("message" to "Bitte eine gültige E-Mail-Adresse eintragen."))
        }
        settingsService.saveMailFromAddress(address)
        return ResponseEntity.ok(
            mapOf(
                "message" to if (address.isBlank()) {
                    "Absender zurückgesetzt – Auto-Mails nutzen wieder den SMTP-Benutzer."
                } else {
                    "Absender für automatische Mails gespeichert."
                },
            ),
        )
    }

    @GetMapping("/anfrage-funnel-spamfilter")
    fun getFunnelSpamFilter(): ResponseEntity<FunnelSpamFilterResponse> {
        return ResponseEntity.ok(FunnelSpamFilterResponse(settingsService.isAnfrageFunnelSpamFilterAktiv))
    }

    @PutMapping("/anfrage-funnel-spamfilter")
    fun saveFunnelSpamFilter(@RequestBody req: FunnelSpamFilterRequest): ResponseEntity<Map<String, String>> {
        settingsService.saveAnfrageFunnelSpamFilterAktiv(req.aktiv)
        return ResponseEntity.ok(
            mapOf(
                "message" to if (req.aktiv) {
                    "Spam-Filter für Webseiten-Anfragen aktiviert."
                } else {
                    "Spam-Filter für Webseiten-Anfragen deaktiviert."
                },
            ),
        )
    }

    private fun hasValue(value: String?): Boolean {
        if (value == null) return false
        val normalized = value.trim().lowercase(Locale.ROOT)
        return normalized.isNotBlank() &&
            normalized != "override_in_local" &&
            normalized != "smtp.example.com" &&
            normalized != "change_me_strong_password"
    }

    data class SmtpSettingsResponse(val host: String, val port: Int, val username: String, val passwordSet: Boolean)
    data class SmtpSettingsRequest(val host: String, val port: Int, val username: String, val password: String?)
    data class SmtpTestRequest(
        val host: String?,
        val port: Int,
        val username: String?,
        val password: String?,
        val testRecipient: String?,
    )
    data class ImapSettingsResponse(val host: String, val port: Int, val username: String, val passwordSet: Boolean)
    data class ImapSettingsRequest(val host: String, val port: Int, val username: String, val password: String?)
    data class ImapTestRequest(val host: String?, val port: Int, val username: String?, val password: String?)
    data class EmailAccountRequest(val email: String, val password: String?)
    data class GeminiSettingsResponse(val apiKeySet: Boolean)
    data class GeminiSettingsRequest(val apiKey: String?)
    data class GeminiTestRequest(val apiKey: String?)
    data class FunnelSpamFilterResponse(val aktiv: Boolean)
    data class FunnelSpamFilterRequest(val aktiv: Boolean)
    data class MailFromResponse(val address: String?, val smtpUsername: String)
    data class MailFromRequest(val address: String?)

    companion object {
        private val EMAIL_PATTERN: Pattern = Pattern.compile("^[^@\\s]+@[^@\\s.]+(?:\\.[^@\\s.]+)+$")
    }
}
